### PR TITLE
feat(GithubMultiOrgReaderProcessor): implement processor to handle multi-github org ingestion

### LINKED ISSUE #4982: Support namespacing groups ingested via the GithubOrgReaderProcessor
## Feature Suggestion

The current functionality for importing GitHub org data allows for targeting a specific GitHub org and ingesting all users/team as `Users`/`Groups` under the `default` namespace.

Under our use-case, we have a GHE instance that supports having multiple GitHub orgs (usually created per dept. or domain). If we import multiple orgs that contain the same team names then they will both merge under the same `Group` in the `default` namespace (I assume, haven't tested). It would be helpful if we could namespace the imported `Group` that indicates the GitHub org they reference

For example, I have the following GitHub orgs and teams: 
FooOrg: team-abc, team-efg
BarOrg: team-abc, team-xyz

After importing from both orgs, I would expect the following entities:
`Group:FooOrg/team-abc`, `Group:FooOrg/team-efg`, `Group:BarOrg/team-abc`, `Group:BarOrg/team-xyz`

This would allow components to specify their `owner` fields to correctly reflect team ownership.

Another thing I would like to note is that users in our GHE may have multi-org memberships so in our use-case it wouldn't make sense to namespace `Users` since there will only be one unique set of usernames.  

## Context
 There might be related to discussions in https://github.com/backstage/backstage/issues/4089 and I'm not sure how exactly this would fit in with the proposal described there.

### PR DESCRIPTION
**Pull request by kuangp**
## Hey, I just made a Pull Request!

Putting up our internal implementation of a GitHub multi-org ingestion processor for feedback.

This processor heavily extends from the `GithubOrgReaderProcessor` to support namespacing groups according to their orgs to prevent potential conflicts. One thing we can do to make it more generic is to expose config options in app-config.yaml to take in a list of orgs and default to ingesting all orgs if undefined. Other ideas include allowing the config options to specify namespace to use per given org eg.:
```
type: github-multi-org
target: https://github.<my-org>.com
orgs:
  - name: fooOrg
    namespace: foo
  - name: barOrg
    namespace: bar
```
We could also potentially enhance`GithubOrgReaderProcessor` with the multi-org support implemented here and preserve current behavior as the default by using the default namespace if only one org is specified instead of creating a separate processor.

Closes #4982
#### :heavy_check_mark: Checklist

- [x] A changeset describing the change and affected packages. ([more info](https://github.com/backstage/backstage/blob/master/CONTRIBUTING.md#creating-changesets))
- [ ] Added or updated documentation
- [ ] Tests for new functionality and regression tests for bug fixes
- [ ] Screenshots attached (for UI changes)
- [x] All your commits have a `Signed-off-by` line in the message. ([more info](https://github.com/backstage/backstage/blob/master/CONTRIBUTING.md#developer-certificate-of-origin))

### CODE REVIEW

**Comment by Rugvip** [comment]
Thanks! Looks good, and we have indeed considered adding more parameters to the locations, both on the entities and on the specs or in config. There's a bit of a debate to be had over whether we should even use catalog processors at all for this kind of integration in the future, as we're exploring some more direct ways to feed the catalog that would potentially make deletions more simple to do. But that's gonna have to wait and it should be relatively simple to migrate code either way.

Regarding the exact model that is being used here, with namespaces used to distinguish between different orgs, that's also a hotly debated topic and https://github.com/backstage/backstage/issues/4089 is the best place to discuss that. We're not quite sure if we want to encourage that pattern just yet or if we want to go with some other approach, like prefixing. Entity names can technically contain `/`'s, so just using the org/user slug could work quite well too.

We could potentially mark this up with a couple more warning signs if you want to work it into something that we're happy to merge? Just making sure that it's clear that it's very unstable API that could potentially change in the future, including the output of the processor.

**Comment by kuangp** [comment]
@Rugvip thanks for the info! 

I am aware that usage of namespaces in this manner may clash with whatever eventually comes out of #4089 but wasn't sure of a better alternative. If we set the entity name to the org/user slug as you suggest, is there a way to reference this correctly in the `owner` field/relations of other components? My understanding was that the prefix before the `/` will be interpreted as a namespace when the `owner` field is parsed which is why the [schema](https://backstage.io/docs/features/software-catalog/descriptor-format#name-required) restricts usage of the `/` char.

That said, I am happy to flush this implementation out to be mergable with the project with the current state of the API. I'll open this PR up for review to welcome additional feedback before I make further changes.

File: plugins/catalog-backend/src/ingestion/processors/GithubMultiOrgReaderProcessor.ts
Comment (adamdmharvey) [inline plugins/catalog-backend/src/ingestion/processors/GithubMultiOrgReaderProcessor.ts:103]:
```suggestion
        `Read ${users.length} GitHub users and ${groups.length} GitHub teams from ${org} in ${duration} seconds`,
```

I know there's nuance here, in that GitHub calls them teams but Backstage groups, so we're essentially trying to ingest teams to MAKE them groups in Backstage, but considering this msg is specific about what it's done IN GitHub, maybe this should say teams?

File: plugins/catalog-backend/src/ingestion/processors/github/github.ts
Comment (adamdmharvey) [inline plugins/catalog-backend/src/ingestion/processors/github/github.ts:73]: I'm no expert here, but I believe the default Octokit library only calls GitHub.com.  Is there anything we need to pass in here for this to support GHE?

**Review by adamdmharvey — COMMENTED** [review]
Excited by this !

File: plugins/catalog-backend/src/ingestion/processors/github/github.ts
Comment (kuangp) [reply plugins/catalog-backend/src/ingestion/processors/github/github.ts:73]: So I think this method should expect the passed in Octokit client to already be configured correctly to point to a GHE instance which I do here: https://github.com/backstage/backstage/pull/5173/files#diff-18d2ea790aeb3068a8e8abc3907e72e37617eef845e0a02bfdf24086d3abca50R60

**Comment by alexcurtin** [comment]
Any movement on this? Definitely excited to have this supported in backstage!

**Comment by kuangp** [comment]
@alexcurtin happy to continue work here but waiting on additional feedback on the direction of this change from @Rugvip and/or core maintainers on best way to proceed here.

File: plugins/catalog-backend/src/ingestion/processors/github/github.ts
Comment (Rugvip) [inline plugins/catalog-backend/src/ingestion/processors/github/github.ts:214]: Could call this `groupKey` just to be a bit more clear with the purpose

**Review by Rugvip — COMMENTED** [review]
Sorry for the long wait, tricky topic and lots going on.

At this point I'm actually thinking we ship this, thoughts on that @backstage/maintainers? I'm coming to the conclusion that this might be a completely fine way to work with GitHub orgs, even if we end up using namespaces for access control or differentiating auth providers that'd just result in that new addition being awkward to use or incompatible with this.

Basically using this processor will give you a catalog model that is compatible with all current usage of the catalog, but maybe not all future usage. It may be that users of the processors would have to migrate ownership in the future if they want to use a specific future feature, like multi-provider sign-in, but it might also be that there would need to be another migration if we chose a different solution than what we have here. As long as we slap a warning label on the processor I'm thinking we're good.

So apart from the nit, please add a changeset and a warning in the TsDoc for the processor. Slap an `@alpha` tag on it and mention that it may not be compatible with future org structures in the catalog

**Comment by freben** [comment]
Agreed. It sounds ok to unblock on this and defer future compat concerns to when and if they happen

**Comment by ltdonner** [comment]
Hello, everyone. Does anybody know if this is still being worked on? It looks like it's been almost a month since there was activity on it but it also sounded like with some small changes this would be okay with everyone to merge?

**Comment by ltdonner** [comment]
@kuangp Thank you for the update, I appreciate it!

**Comment by kuangp** [comment]
@Rugvip @freben flushed out this change so it's more generic/compatible with upstream. Still need to add/update unit tests but wanted to get your feedback on the overall structure of the change

File: packages/integration/src/github/GithubCredentialsProvider.ts
Comment (kuangp) [inline packages/integration/src/github/GithubCredentialsProvider.ts:126]: Wasn't sure the best way to handle etags here for a paginated request. Currently this looks like it only fetched the first 30 installations according to the [docs](https://docs.github.com/en/rest/reference/apps#list-installations-for-the-authenticated-app). I updated this to use `paginate` to ensure this scales to fetch all installations but this doesn't give us an etag for us to track

**Comment by alexcurtin** [comment]
This is eagerly awaited, thanks so much!

**Comment by kuangp** [comment]
@Rugvip @freben for another look when you get a chance, thanks!

File: packages/integration/src/github/GithubCredentialsProvider.ts
Comment (freben) [inline packages/integration/src/github/GithubCredentialsProvider.ts:130]: I realize this isn't part of your pr, but it feels to me that maybe this querying should be cached and only done at most once per X time unit. If I read things correctly, we reach out to the github api over and over again, only to grab a single one out of the list, and then the caching logic comes after that.

Well, something for future improvement perhaps

File: plugins/catalog-backend/src/ingestion/processors/github/github.ts
Comment (freben) [inline plugins/catalog-backend/src/ingestion/processors/github/github.ts:211]: Hm, yeah I see what you are doing here. It's a bit unfortunate that the key here is a string, since if you defined that your user is `memberOf: ['group:someNamespace/someName']` (i.e. over-specifying the ref to include the optional kind as well), this won't match the string `someNamespace/someName` that comes out of this.

Not a big deal though. Let's keep this and then consider refactoring to use a full `EntityName` always instead, or similar

File: plugins/catalog-backend/src/ingestion/processors/GithubMultiOrgReaderProcessor.ts
Comment (freben) [inline plugins/catalog-backend/src/ingestion/processors/GithubMultiOrgReaderProcessor.ts:156]: Thinking about the error cases here, maybe it's a bit odd that you emit the users here outside the catch but groups are treated differently?

Maybe only wrap the things that can actually fail (like the `getOrganization*` functions) in a try/catch and abort entirely if that fails? And then always wire up and emit both users and groups if it didn't fail

**Review by freben — APPROVED** [review]
Small comments but no blockers

File: plugins/catalog-backend/src/ingestion/processors/GithubMultiOrgReaderProcessor.ts
Comment (kuangp) [reply plugins/catalog-backend/src/ingestion/processors/GithubMultiOrgReaderProcessor.ts:156]: My thinking here was that if processing a particular GitHub org fails for any reason, it would log the failure but continue to process the remaining orgs which is why I wrapped the body of the `for` loop in a try/catch to continue the iteration in case of errors. 

As it is currently neither any users or groups from the particular org would be emitted if either of the `getOrgranization*` functions throws since `allUsersMap` would not be updated.

**Comment by kuangp** [comment]
@Rugvip any comments/concerns before this is mergeable?

**Review by freben — APPROVED** [review]

**Pull request merged**
